src/ui/public/state_management/state_monitor_factory.js

```diff
@@ -37,9 +37,11 @@ function stateMonitor(state, customInitialState) {
 
   function dispatchChange(type = null, keys = []) {
     const status = getStatus();
-    changeHandlers.forEach(changeHandler => {
-      changeHandler(status, type, keys);
-    });
+    if (changeHandlers) {
```

Comment:
Without this change, an error is thrown when deleting a panel.

I can resolve the issue another way - by swapping the order of dispatches in onDeletePanel to first destroyEmbeddable and then deletePanel, but this feels like a fragile fix.  Really the order shouldn't matter.  Here is the issue with deleting the panel first:
 
  - DeletePanel is dispatched
  - This triggers _handleStoreChanges to be called inside of `dashboard_state_manager`.
  - This updates the panels array and calls saveState so appState updates the url accordingly.
  - saveState emits `save_with_changes` which the state monitor is listening to.
  - next destroyEmbeddale is dispatched.
  - this destroys the visualization DOM node, which in turn destroys a state monitor that was listening to `save_with_changes`
   - The next event loop starts and  `save_with_changes` is triggered, which causes this to be called, but because destroy was already called, it throws an error.

If destroyEmbeddable is called first, this destroys the dom node and turns off the state monitor listeners before `this.saveState` emits the change event.

But the real problem is not that the order is wrong, but that off'ing an event emission won't stop that state monitor from reacting to already emitted events.  Here's some debug output to show the problem: (I gave state monitor instances ids to track them)

<img width="444" alt="screen shot 2017-10-13 at 10 38 42 pm" src="https://user-images.githubusercontent.com/16563603/31571799-5c309924-b067-11e7-8070-a4fe0ed59cf5.png">

... I actually went ahead and filed https://github.com/elastic/kibana/issues/14462.  It probably won't be too difficult to fix the root problem and then I can remove this extra check.

Reply:
Good catch!